### PR TITLE
RUF010: New rule for removing blank comments

### PR DESCRIPTION
**Pull request by andreyfedoseev**
Feature proposal: https://github.com/charliermarsh/ruff/issues/4238

### CODE REVIEW

**Comment by charliermarsh** [comment]
Thanks for this! Looking through the codebases, I think most of the errors flagged in the ecosystem check are false positives. Could we try implementing this by looking at the token stream, rather than using a regex on the physical lines? (A regex will also be more expensive in practice.)

E.g., in the `tokens.rs` checker, we could look at `Comment` tokens, see if they're blank, and track some state to see if they're part of a multi-line comment. Would that work?

**Comment by MichaReiser** [comment]
> Could we try implementing this by looking at the token stream, rather than using a regex on the physical lines? (A regex will also be more expensive in practice.)

It should be possible to use `Indexer.comment_ranges` to avoid iterating over all tokens.

**Comment by charliermarsh** [comment]
As an aside, it looks like this can be implemented as `PLR2044`: https://pylint.readthedocs.io/en/latest/user_guide/messages/refactor/empty-comment.html

File: crates/ruff/src/rules/ruff/rules/blank_comment.rs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+static BLACK_COMMENT_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r"\S(\s*#\s*)$").unwrap());
+
+/// RUF010
+pub fn blank_comment(
+    diagnostics: &mut Vec<Diagnostic>,
+    line: &Line,
+    settings: &Settings,
+    autofix: flags::Autofix,
+) {
+    if let Some(captures) = BLACK_COMMENT_REGEX.captures(line.as_str()) {
```
Comment (MichaReiser) [inline crates/ruff/src/rules/ruff/rules/blank_comment.rs:51]: Ruff has the `Indexer` data structure that exposes the range of all comments (`Indexer::comment_ranges`). You can use this with `Locator.slice(range)` to retrieve all comments and call the rule at the top level (module) once. This should be faster than running a regex on every line. 

I think we should also be able to replace the regex all together by manually skipping the `#` and then testing if `text.trim().is_empty()`.

File: crates/ruff/src/rules/ruff/snapshots/ruff__rules__ruff__tests__RUF010_RUF010.py.snap [hunk 2]
```diff
@@ -0,0 +1,42 @@
+---
+source: crates/ruff/src/rules/ruff/mod.rs
+---
+RUF010.py:6:23: RUF010 [*] Blank comments are useless and should be removed
+  |
+6 | # is valid
+7 | 
+8 | print("Hello, world!")  #
+  |                       ^^^ RUF010
```
Comment (MichaReiser) [inline crates/ruff/src/rules/ruff/snapshots/ruff__rules__ruff__tests__RUF010_RUF010.py.snap:9]: Nit: Can we change the range to only highlight the comment

File: crates/ruff/src/rules/ruff/rules/blank_comment.rs (hunk 1 above)
Comment (sladyn98) [reply crates/ruff/src/rules/ruff/rules/blank_comment.rs:51]: Yeah I guess we could do something like this

```
// Retrieve the comment using Locator.slice
        let comment = Locator.slice(range);

        // Skip the # at the start of the comment and trim white spaces
        let trimmed_comment = comment.strip_prefix("#").unwrap_or("").trim();

        // If the trimmed comment is empty, it's a blank comment
        if trimmed_comment.is_empty() {
            let mut diagnostic = Diagnostic::new(BlankComment, range);
            if autofix.into() && settings.rules.should_fix(Rule::BlankComment) {
                diagnostic.set_fix(Edit::deletion(range.start(), range.end()));
            }
            diagnostics.push(diagnostic);
        }
    }
    
 ```

**Comment by andreyfedoseev** [comment]
Below are a few examples that make it somewhat tricky:

1. Basic case, blank comment follows some code, comment should be removed

```
print("hello")  #
```

2. Multi-line comments, blank lines should be preserved:

```
# First line
# 
#
# Last line
# 
```

3. However, blank multi-line comments should be removed:

```
# The blank comment below should be removed

#
#

#
# This should not be removed
#

```

4. Also, we need to be careful about what we consider a multi line comment

```
# The comment on the next line is not part of a multi-line comment
print("hello")  #
```

For that we need to keep track two types of context for each comment token:

1. Does a comment token follow a code fragment?
2. Is a comment token a part of a continuous multi-line comment?

**Comment by sladyn98** [comment]
We could use a commentContext for something like this
```
pub struct CommentContext {
    is_comment_continuous: bool,
    is_comment_after_code: bool,
}

pub fn check_blank_comment(
    diagnostics: &mut Vec<Diagnostic>,
    line: &Line,
    settings: &Settings,
    autofix: flags::Autofix,
    context: &mut CommentContext,
) {
    // Retrieve the comment using Locator.slice
    let comment = Locator.slice(range);

    // Skip the # at the start of the comment and trim white spaces
    let trimmed_comment = comment.strip_prefix("#").unwrap_or("").trim();

    // Flag to check if current comment is blank
    let is_blank = trimmed_comment.is_empty();

    // If the trimmed comment is empty and either not part of continuous comment or after a code, it's a blank comment to be removed
    if is_blank && (!context.is_comment_continuous || context.is_comment_after_code) {
        let mut diagnostic = Diagnostic::new(BlankComment, range);
        if autofix.into() && settings.rules.should_fix(Rule::BlankComment) {
            diagnostic.set_fix(Edit::deletion(range.start(), range.end()));
        }
        diagnostics.push(diagnostic);
    }

    // Update the context for the next line
    context.is_comment_continuous = !is_blank;
    context.is_comment_after_code = false;
}

// You'll need to handle these two context flags while parsing the whole script.

```

**Comment by MichaReiser** [comment]
@andreyfedoseev I think the todo rule does something very similar. It checks whether two comments are in a row and if it contains the necessary metadata

https://github.com/charliermarsh/ruff/blob/c6e5fed6587c95839a814ee8dfa845b465351229/crates/ruff/src/rules/flake8_todos/rules.rs#L293-L339

**Comment by charliermarsh** [comment]
Gonna close this out for now since it's been inactive for a bit -- aways happy to re-open if we decide to pick the work back up :)

**Pull request closed without merging**